runtime/gateway.go

```diff
@@ -909,17 +906,29 @@ func GetHostname() string {
 	return host
 }
 
-// shutdownTChannelServer gracefully shuts down the tchannel server, blocks until the shutdown is
+// shutdownTChannelServerAndClients gracefully shuts down the tchannel server, blocks until the shutdown is
 // complete or the timeout has reached if there is one associated with the given context
-func (gateway *Gateway) shutdownTChannelServer(ctx context.Context) error {
+// It also shuts down all the dedicated client tchannel connections on a best effort basis
+func (gateway *Gateway) shutdownTChannelServerAndClients(ctx context.Context) error {
 	shutdownPollInterval := defaultShutdownPollInterval
 	if gateway.Config.ContainsKey("shutdown.pollInterval") {
 		shutdownPollInterval = time.Duration(gateway.Config.MustGetInt("shutdown.pollInterval")) * time.Millisecond
 	}
 	ticker := time.NewTicker(shutdownPollInterval)
 	defer ticker.Stop()
 
+	for serviceName, clientTchannel := range gateway.ClientTChannels {
+		go func(service string, ch *tchannel.Channel) {
+			gateway.Logger.Info(fmt.Sprintf("Closing tchannel client for [%v]", service))
+			ch.Close()
```

Comment:
This could lead to a situation where client is closed but server is not? Leading to failures in pending outbound calls during shtudown (similar to https://github.com/uber/zanzibar/pull/764).

Can we close all the servers first, wait for their server.close() to complete, and then shutdown clients?

Reply:
+1

Reply:
Good catch. Updating